src/webui/api/torrentscontroller.cpp

```diff
@@ -252,6 +254,23 @@ namespace
             idList << BitTorrent::TorrentID::fromString(hash);
         return idList;
     }
+
+    nonstd::expected<QUrl, QString> validateWebseedUrl(const QString &urlStr)
+    {
+        // sanity check before parsing to prevent QUrl from checking for the existence of files on the filesystem
+        if (!supportedWebseedSchemes.contains(urlStr.section(u"://"_s, 0, 0)))
+            return nonstd::make_unexpected(TorrentsController::tr("URL scheme must be one of [%1]").arg(supportedWebseedSchemes.join(u",")));
+
+        const QUrl url(urlStr, QUrl::StrictMode);
+        if (!url.isValid())
+            return nonstd::make_unexpected(TorrentsController::tr("\"%1\" is not a valid url").arg(urlStr));
+
+        // sanity check after parsing in case QUrl interpreted a different scheme than intended
```

Comment:
I'm not sure if it is, but QUrl is a blackbox that we have very little control over. It seemed worthwhile to have a sanity check after parsing.

Reply:
> I'm not sure if it is, but QUrl is a blackbox that we have very little control over. It seemed worthwhile to have a sanity check after parsing.

There is a subject area. If `QUrl` returns an unexpected scheme, then you need to report this obvious error to Qt, and not hide it in the code behind such sanity checks.

Reply:
Sure, but that strategy doesn't actually help us in practice. This current approach is far more pragmatic and adds no iverhead

Reply:
> This current approach is far more pragmatic

I disagree. Just try to reproduce the operation of this function, provided that the incoming URL is valid (like `https://example.com/announce`) in the case when `QUrl` returns the expected scheme (`https`), and when it does not. In the second case, your "pragmatic approach" will return some nonsense as an error message.

Reply:
On this point we disagree. I see no harm in keeping the check and consider it a best practice, so I will be leaving it in.

Reply:
>On this point we disagree.

I repeat the above in other words:
1. If the `QUrl` is correctly implemented (and we usually assume that this is the case when we use something), then discussed code piece is completely meaningless given the similar check made above.
2. If you assume that `QUrl` may not function correctly, then discussed code piece still does not bring any benefit, but only makes even more confusion by producing inappropriate error message.

>I see no harm in keeping the check and consider it a best practice, so I will be leaving it in.

I can hardly call it the "best practice" to leave deliberately incorrect/useless code.

@Chocobo1 
I wonder if you just overlooked it, or do you also consider it "best practice"?

Reply:
> I can hardly call it the "best practice" to leave deliberately incorrect/useless code.

The best practice I was referring to was validating that the output produced by an external library matches the expected output, especially when there are potential security implications to that output's use.

This is a really strange thing to spend time arguing on. Can we focus on more substantive issues?

Reply:
> The best practice I was referring to was validating that the output produced by an external library matches the expected output, 

This is not applicable in the general case, otherwise it would make the use of external libraries meaningless.
>especially when there are potential security implications to that output's use.

Even if such validation makes sense in some particular case, at least it should not distort the main logic, so it should output a suitable error message, e.g. "The operation failed due to unexpected behavior of QUrl::scheme. Please file a bug report to the developers."

> This is a really strange thing to spend time arguing on. Can we focus on more substantive issues?

Why should we ignore any of the detected flaws in the code being reviewed?
I have no desire to poke around in this, deliberately looking for the slightest reason to slow down work. But this is something that lies on the surface, so I stumbled over it.

Reply:
> Even if such validation makes sense in some particular case, at least it should not distort the main logic, so it should output a suitable error message, e.g. "The operation failed due to unexpected behavior of QUrl::scheme. Please file a bug report to the developers."

The error message is user facing, which would not be an appropriate place for this. What you're describing is best suited for `qDebug`.

Reply:
> The error message is user facing, which would not be an appropriate place for this.

Why? It reports an error. In addition, it gives an explicit indication of the fact that the error is not in the URL specified by the user, but in the incorrect behavior of the application component, which is desirable to inform the developers.
In any case, it's better than reporting some nonsense, like "URL scheme must be one of http, https, ftp" when user has provided "https://example.com" URL.
You can use other words, but the error message should match the error, right?

Reply:
This is an API meant for consumers - not an internal API meant for qBittorrent developers.

> You can use other words, but the error message should match the error, right?

Yes. In this case, the issue is that the interpreted scheme is not one of the allowed schemes. Surely this wouldn't happen with a well formed URL, so the error message is accurate.

Reply:
https://github.com/qbittorrent/qBittorrent/pull/21043#discussion_r1706397004

Reply:
> Surely this wouldn't happen with a well formed URL, so the error message is accurate.

But in the discussed code (i.e. before latest changes) this error message was produced under branch of well formed URL (i.e. after you have made sure that it has the correct scheme and that the resulting QUrl is valid). 

In the latest version of the code, this is no longer relevant, so you can stay with your opinion.